src/oci-umount.c

```diff
@@ -316,6 +335,50 @@ static int map_mount_host_to_container(const struct config_mount_info *config_mo
 	return nr_mapped;
 }
 
+/* Returns 0 on success, negative error otherwise */
+static int unmount(char *umount_path, bool submounts_only, const struct mount_info *mnt_table, size_t table_sz)
+{
+	int ret;
+	unsigned i;
+	unsigned mntid = 0;
+	bool found_mnt = false;
+
+	if (!submounts_only) {
+		ret = umount2(umount_path, MNT_DETACH);
+		if (!ret)
+			pr_pinfo("Unmounted: [%s]\n", umount_path);
+		else
+			pr_perror("Failed to unmount: [%s]\n", umount_path);
+		return ret;
+	}
+
+	/* Unmount submounts only */
+	for (i = 0; i < table_sz; i++) {
+		if (!strcmp(umount_path, mnt_table[i].destination)) {
+			found_mnt = true;
+			mntid = mnt_table[i].mntid;
+			break;
+		}
+	}
+
+	if (!found_mnt) {
+		pr_perror("Could not determine mount id of mountpoint: [%s]\n", umount_path);
+		return -1;
+	}
+
+	/* lazy unmount all direct submounts */
+	for (i = 0; i < table_sz; i++) {
+		if (mnt_table[i].parent_mntid != mntid)
+			continue;
+		ret = umount2(mnt_table[i].destination, MNT_DETACH);
+		if (!ret)
+			pr_pinfo("Unmounted submount: [%s]\n", mnt_table[i].destination);
+		else
+			pr_perror("Failed to unmount submount: [%s]. Skipping.\n", mnt_table[i].destination);
```

Comment:
In most of the cases mount point should be mounted. We have got mount point information from /proc/self/mouninfo. So this should hit rarely and if it hits, it will give us an idea of a potential problem or wrong assumption etc. So I would rather keep it this way.